ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModelTest.kt

```diff
@@ -92,13 +105,159 @@ class PaywallViewModelTest {
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+        )
+
+        model.awaitPurchaseSelectedPackage(activity)
+
+        coVerify { myAppPurchaseLogic.performPurchase(any(), any()) }
+
+        verifyOrder {
+            listener.onPurchaseStarted(any())
+            listener.onPurchaseCompleted(customerInfo, any())
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+        assertThat(dismissInvoked).isTrue
+    }
+
+    @Test
+    fun `Calls purchase cancel listener when purchasesAreCompletedBy == MY_APP`() = runTest {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+
+        val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)
+
+        coEvery { myAppPurchaseLogic.performPurchase(any(), any()) } returns MyAppPurchaseResult.Cancellation
+
+        val model = create(
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+        )
+
+        model.awaitPurchaseSelectedPackage(activity)
+
+        coVerify { myAppPurchaseLogic.performPurchase(any(), any()) }
+
+        verifyOrder {
+            listener.onPurchaseStarted(any())
+            listener.onPurchaseCancelled()
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+        assertThat(dismissInvoked).isTrue
+    }
+
+    @Test
+    fun `Calls purchase error listener when purchasesAreCompletedBy == MY_APP`() = runTest {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+
+        val myAppPurchaseLogic = mockk<MyAppPurchaseLogic>(relaxed = true)
+        val notAllowedError = PurchasesError(PurchasesErrorCode.PurchaseNotAllowedError)
+
+        coEvery { myAppPurchaseLogic.performPurchase(any(), any()) } returns MyAppPurchaseResult.Error(notAllowedError)
+
+        val model = create(
+            customPurchaseLogic = myAppPurchaseLogic,
+            activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+        )
+
+        model.awaitPurchaseSelectedPackage(activity)
+
+        coVerify { myAppPurchaseLogic.performPurchase(any(), any()) }
+
+        verifyOrder {
+            listener.onPurchaseStarted(any())
+            listener.onPurchaseError(notAllowedError)
+        }
+
+        assertThat(model.actionInProgress.value).isFalse
+    }
+
+    @Test
+    fun `Cannot create PaywallViewModel when purchasesAreCompletedBy == MY_APP without custom purchase logic`() {
+        every { purchases.purchasesAreCompletedBy } returns PurchasesAreCompletedBy.MY_APP
+
+        try {
+            create(
+                activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
+            )
+            fail("Expected IllegalStateException to be thrown")
+        } catch (e: IllegalStateException) {
+            // Success: caught the expected exception
+        }
```

Comment:
I would either add `@Suppress("SwallowedException")` right above the `try` to satisfy Detekt, or if you want to go all-in on AssertJ, you can do: 
```suggestion
        assertThatExceptionOfType(IllegalStateException::class.java).isThrownBy {
            create(
                activeSubscriptions = setOf(TestData.Packages.weekly.product.id)
            )
        }
```